Adding Event timeline track

wilaw opened the pull request:
This PR renames the timeline track to "Media timeline", changes its definition so that it refers only to the media published by the original publisher and then adds a new Event timeline track defintion to hold externally generated timeline data. 

Fixes #66

wilaw committed:
Adding Event timeline track

gwendalsimon commented on draft-ietf-moq-warp.md:
```diff
@@ -953,23 +955,22 @@ Each subsequent Group ID MUST increase by 1.
+track MUST be specified as "text/csv" in the catalog.
 
-Each timeline track begins with a header row of GROUP_ID,OBJECT_ID,MEDIA_PTS,
+Each media timeline track begins with a header row of GROUP_ID, OBJECT_ID, MEDIA_PTS,
```
Location is now formally defined in MOQT spec https://www.ietf.org/archive/id/draft-ietf-moq-transport-14.html#section-1.4.1 so, instead of having GROUP_ID and OBJECT_ID in the timeline, we could prefer LOCATION. If the formatting of Location changes in the future, the media timeline would stay valid.

wilaw replied:
Commit https://github.com/moq-wg/warp-streaming-format/pull/67/commits/d4ad651876505e5ffe2e797da6572977f81a04b4 switches to using Location

gwendalsimon commented on draft-ietf-moq-warp.md:
```diff
@@ -953,23 +955,22 @@ Each subsequent Group ID MUST increase by 1.
+The media timeline track provides data about the previously published groups and their
+relationship to wallclock time and media time. Timeline tracks allow players to seek to
+precise points behind the live head in a live broadcast, or for random access in a
+VOD asset. The optional metadata can describe a characteristic of any record in the
```
Should we keep a metadata field in the Media Timeline? 
In my understanding, the media timeline provides all the info for seeking while any additional information should be given in the event timeline.
Do you have some examples of metadata that could be necessary to include in the media timeline?

wilaw replied:
Yes, I thought we could remove it too, however @yekuiwang  had a use-case, of describing the type of SAP access point represented by the Group. So I think there is some valid metadata that can describe the produced content. In many cases, it may be empty.

gwendalsimon commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,91 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+cars, or active speaker notifications in web conferences.
+
+To allow the client to bind this event metadata with the broadcast content described by
+the media timeline track, each event record MUST contain a reference to either GroupID,
```
Location instead of GroupID?

wilaw replied:
Location defines a binary serialization. We have a JSON doc here. Do you propose somehting like this? 

> [
>     {
>         "location": { "g":432, "o": 3},
>         "type": "com.example.custom.live-score-data",
>         "data": {
> ...

or can we do an array

> [
>     {
>         "location": [432, 3],
>         "type": "com.example.custom.live-score-data",
>         "data": {
> ...

Note that location doesn't allow the Object ID to be undefined or have a default value, so if its used then both groupID and object ID must be supplied.

wilaw replied:
Commit https://github.com/moq-wg/warp-streaming-format/pull/67/commits/d4ad651876505e5ffe2e797da6572977f81a04b4 defines the JSON representation of Location using an array of numbers.

gwendalsimon commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,91 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+cars, or active speaker notifications in web conferences.
+
+To allow the client to bind this event metadata with the broadcast content described by
+the media timeline track, each event record MUST contain a reference to either GroupID,
```
(english writing) Can we say "either" when we have three element? It is vague.

We should prefer:
- only one of three identifiers (Location, Media_PTS, Wallclock time)
- any of the three identifiers  ==> In that case, the formatting may become a problem and we may need to write some sentences about possible inconsistencies.

I would be in favor of having only one of the three.

wilaw replied:
I'm fine with one. Simplicity is  always the better choice.

wilaw replied:
Commit https://github.com/moq-wg/warp-streaming-format/pull/67/commits/157d5d655fdf6bc2be1a2c5e5354bf0b5fa95c9a restricts the record to only supplying one identifier.

gwendalsimon commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,91 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+
+* A timeline reference, which MUST be either 't' for wallclock time, 'g' for Group ID or
+  'm' for Media PTS. Typically only one timeline reference is included in each record,
+  although multiple timeline entries may exist within a record, as long as they
```
Yes, I can see my comment https://github.com/moq-wg/warp-streaming-format/pull/67/files#r2362812236 is addressed here. So you may ignore. I keep it in case we want to discuss it.

gwendalsimon reviewed (commented):
Thank you for the PR. I added minor comments, but overall it is good to me.

wilaw committed:
Modifying rules for timeline index types

Only one of 't' for wallclock time, 'g' for Group ID or
  'm' for Media PTS is allowed per record.

yekuiwang commented on draft-ietf-moq-warp.md:
```diff
@@ -953,23 +955,22 @@ Each subsequent Group ID MUST increase by 1.
+precise points behind the live head in a live broadcast, or for random access in a
+VOD asset. The optional metadata can describe a characteristic of any record in the
+timeline. Timeline tracks are optional. Multiple timeline tracks can exist inside a
+catalog.
```
```suggestion
relationship to wallclock time and media time. Media timeline tracks allow players to seek to
precise points behind the live head in a live broadcast, or for random access in a
VOD asset. The optional metadata can describe a characteristic of any record in the media 
timeline. Media timeline tracks are optional. Multiple media timeline tracks can exist inside a
catalog.

yekuiwang commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,90 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+of "mediatimeline". A media timeline track MUST carry a 'depends' attribute which
+contains an array of all track names to which the media timeline track applies.
+
+## Media Timeline track updating.
```
```suggestion
## Media Timeline track updating

yekuiwang commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,90 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+## Media Timeline track updating.
+The publisher MUST publish an indepdendent media timeline in the first MOQT Object
+of each MOQT Group of a media timeline track. The publisher MAY publish incremental
+updates in the second and subsequent Objects within each GROUP. Incremental updates
```
```suggestion
updates in the second and subsequent Objects within each Group. Incremental updates

yekuiwang commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,90 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+
+To allow the client to bind this event metadata with the broadcast content described by
+the media timeline track, each event record MUST contain a reference to either GroupID,
+Media PTS or wallclock time.
```
```suggestion
the media timeline track, each event record MUST contain a reference to GroupID,
Media PTS or wallclock time.

yekuiwang commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,90 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+## Event Timeline track updating.
+The publisher MUST publish an indepdendent event timeline in the first MOQT Object
+of each MOQT Group of an event timeline track. The publisher MAY publish incremental
+updates in the second and subsequent Objects within each GROUP. Incremental updates
```
```suggestion
updates in the second and subsequent Objects within each Group. Incremental updates

wilaw committed:
Update draft-ietf-moq-warp.md

Co-authored-by: Ye-Kui Wang <98077873+yekuiwang@users.noreply.github.com>

wilaw committed:
Update draft-ietf-moq-warp.md

Co-authored-by: Ye-Kui Wang <98077873+yekuiwang@users.noreply.github.com>

wilaw committed:
Update draft-ietf-moq-warp.md

Co-authored-by: Ye-Kui Wang <98077873+yekuiwang@users.noreply.github.com>

wilaw committed:
Update draft-ietf-moq-warp.md

Co-authored-by: Ye-Kui Wang <98077873+yekuiwang@users.noreply.github.com>

wilaw committed:
Update draft-ietf-moq-warp.md

Co-authored-by: Ye-Kui Wang <98077873+yekuiwang@users.noreply.github.com>

yekuiwang commented on draft-ietf-moq-warp.md:
```diff
@@ -991,20 +992,90 @@ WALLCLOCK,METADATA. This row defines the 5 columns of data within each record.
+of "eventtimeline". An event timeline track MUST carry a 'depends' attribute which
+contains an array of all track names to which the event timeline track applies.
+
+## Event Timeline track updating.
```
```suggestion
## Event Timeline track updating

wilaw committed:
Fix linting issue

wilaw committed:
Update draft-ietf-moq-warp.md

Co-authored-by: Ye-Kui Wang <98077873+yekuiwang@users.noreply.github.com>

wilaw committed:
Adding support for Location + new example
